horovod/run/run.py

```diff
@@ -0,0 +1,436 @@
+
+import argparse
+import hashlib
+import os
+import sys
+import six
+try:
+    from shlex import quote
+except ImportError:
+    from pipes import quote
+
+import horovod
+from horovod.run.common.util import codec, safe_shell_exec, timeout, secret
+from horovod.run.driver import driver_service
+from horovod.run.task import task_service
+from horovod.run.util import cache, threads, network
+
+# Cached information of horovodrun functions be stored in this directory
+CACHE_FOLDER = os.path.join(os.path.expanduser('~'), '.horovod')
+
+# Cache entries will be stale if they are older than this number of minutes
+CACHE_STALENESS_THRESHOLD_MINUTES = 60
+
+# Number of retries for sshing into the hosts
+SSH_RETRIES = 5
+
+
+@cache.use_cache()
+def _check_all_hosts_ssh_successful(host_addresses, ssh_port=None):
+    """
+    checks if ssh can successfully be performed to all the hosts.
+    :param host_addresses: list of addresses to ssh into. for example,
+        ['worker-0','worker-1']
+        ['10.11.11.11', '10.11.11.12']
+    :type host_addresses: list(strings)
+    :return: Returns True if all ssh was successful into all the addresses.
+    """
+
+    def exec_command(command):
+        exit_code = 1
+        output_msg = ""
+
+        # Try ssh 5 times
+        for i in range(SSH_RETRIES):
+            output = six.StringIO()
+            try:
+                exit_code = safe_shell_exec.execute(command,
+                                                    stdout=output,
+                                                    stderr=output)
+                if exit_code == 0:
+                    break
+                else:
+                    output_msg = output.getvalue()
+            finally:
+                output.close()
+        return exit_code, output_msg
+
+    if ssh_port:
+        ssh_port_arg = "-p {ssh_port}".format(ssh_port=ssh_port)
+    else:
+        ssh_port_arg = ""
+
+    ssh_command_format = 'ssh -o StrictHostKeyChecking=no {host} {ssh_port_arg} date'
+
+    args_list = [[ssh_command_format.format(host=host_address,
+                                            ssh_port_arg=ssh_port_arg)]
+                 for host_address in host_addresses]
+    ssh_exit_codes = \
+        threads.execute_function_multithreaded(exec_command,
+                                               args_list)
+
+    ssh_successful_to_all_hosts = True
+    for index, ssh_status in six.iteritems(ssh_exit_codes):
+        exit_code, output_msg = ssh_status[0], ssh_status[1]
+        if exit_code != 0:
+            print("ssh not successful for host {host}:\n{msg_output}".format(
+                host=host_addresses[index],
+                msg_output=output_msg
+            ))
+
+            ssh_successful_to_all_hosts = False
+    if not ssh_successful_to_all_hosts:
+        exit(1)
+    return True
+
+
+def _launch_task_servers(host_names, local_host_names, driver_addresses,
```

Comment:
Rename `host_names` to `all_host_names` for clarity?

Reply:
done.